engine/render/renderer/framebuffer/Framebuffer.h

```diff
@@ -0,0 +1,219 @@
+        {
+            return depthAttachments;
+        }
+
+        /**
+         * @brief Returns the number of attachments provided to the Config.
+         * @return the number of attachments.
+         */
+        const u32& GetAttachmentCount() const
+        {
+            return attachmentCount;
+        }
+
+        /**
+         * @brief Returns the framebuffer's rendering area width.
+         * @return the rendering area's width.
+         */
+        const u32& GetWidth() const
+        {
+            return width;
+        }
+
+        /**
+         * @brief Returns the framebuffer's rendering area height.
+         * @return the rendering area's height.
+         */
+        const u32& GetHeight() const
+        {
+            return height;
+        }
+
+        /**
+         * @brief Gets the number of configured layers.
+         * @return the number of layers to render to.
+         */
+        const u32& GetLayers() const
+        {
+            return layers;
+        }
+
+    private:
+
+        VkRenderPass renderPass {VK_NULL_HANDLE};
+        FrameImages* colorAttachments {nullptr};
+        FrameImages* depthAttachments {nullptr};
+        u32 attachmentCount {0};
+        u32 width {0};
+        u32 height {0};
+        u32 layers {0};
+    };
+
+    /**
+     * @brief Sets the value of IMAGE_COUNT. This enforces that the number of allocated framebuffers
+     * never exceeds the provided number of images.
+     * @param imageCount the number of images available to the swapchain.
+     */
+    static void SetImageCount(const u32& imageCount)
+    {
+        IMAGE_COUNT = imageCount;
+    }
+
+    /**
+     * @brief A default empty constructor for the Framebuffer.
+     */
+    Framebuffer() = default;
+
+    /**
+     * @brief A constructor for Framebuffer which allows the Framebuffer to be initialised using a
+     * Config object.
+     * @param config
+     * @param device
+     */
+    explicit Framebuffer(const Config& config, const VkDevice& device);
+
+    /**
+     * @brief A default empty destructor.
+     */
+    ~Framebuffer() = default;
+
+    /**
+     * @brief Initialises the Framebuffer using the Configs and Device objects.
+     * @param config the configuration data which will be used to create the Framebuffer.
+     * @param device the device used to allocate the framebuffer.
+     */
+    void Initialise(const Config& config, const VkDevice& device);
+
+    /**
+     * @brief Returns a VkFramebuffer object corresponding to the provided index.
+     * @param index the frame index for the framebuffer.
+     * @return the corresponding VkFramebuffer.
+     */
+    VkFramebuffer GetFramebuffer(const u32& index)
+    {
+        return framebuffers[index];
+    }
+
+    /**
+     * @brief Deallocates the memory used by the framebuffers.
+     * @param device the GPU used to allocate this data.
+     */
+    void DestroyFramebuffer(const VkDevice& device);
```

Comment:
I thought we ditched this convention?

Reply:
I don't believe it's possible for us to not use this convention. Vulkan dynamically allocates data and we need a way to pass it around without it being destroyed every time we copy some data. As nice as RAII is, it causes a ton of issues, especially when it comes to Vulkan.

Reply:
Of course, if you'd LIKE to deallocate every time we destroy these objects then I guess it's possible, I just really don't think it's a good idea from a performance point of view. Ideally we want to create our Vulkan concepts once at the start of the engine and then re-use them as we go. Another option is to use a flag so that we can move some data around without deleting it on the destructor. I just think that's a lot messier and hides information from the user.

Reply:
Not sure what constraints make this the case, but it's something we can discuss at some point.

Reply:
This has been changed. I'll be going over a bunch of data structures in the coming days and re-doing the way we de-allocate memory so that they work with constructors.